compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java

```diff
@@ -1161,11 +1168,23 @@ private BType getExpectedTypeForInferredTypedescMember(BUnionType originalType,
             return null;
         }
 
-        if (types.size() == 1) {
-            return types.iterator().next();
+        // Add the original union type if all the members of the original type are present in the `types` list.
+        LinkedHashSet<BType> originalTypesSet = new LinkedHashSet<>(); // This is to maintain the order of members
+        for (BType originalMemberType : expUnionType.getOriginalMemberTypes()) {
+            LinkedHashSet<BType> flatTypeSet = toFlatTypeSet(new LinkedHashSet<>(Set.of(originalMemberType)));
+            Set<BType> typesToAdd = flatTypeSet.stream().filter(types::contains).collect(Collectors.toSet());
+            if (typesToAdd.containsAll(flatTypeSet)) {
+                originalTypesSet.add(originalMemberType);
+            } else {
+                originalTypesSet.addAll(typesToAdd);
```

Comment:
Is it possible for `types` to not have all the members that were in `expUnionType`? If that's not possible, this is OK, but if it is possible, is this correct? Dropping some types here?

Reply:
Since we are iterating the original member type list we wont me missing any types

Reply:
`typesToAdd` may not have all the members from `flatTypeSet` since we only collect those that are in `types`, right?

Reply:
Consider
```bal
type  A int|string;

type B float|error;

A|B var = dependentlyTypedFunc which returns t|error;
```

For the above case `types` will be `int|string|float`. When we iterate the original types, `A ` will be added by going into the if block. `B` will not be added since `error` is not in the types set as expected. But the `float` will be added by going into the else block

Reply:
OK since we are filtering what's in types.

Reply:
Yes